File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/descriptors/Rowtime.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.descriptors

import org.apache.flink.table.api.Types
import org.apache.flink.table.descriptors.RowtimeValidator.{normalizeTimestampExtractor, normalizeWatermarkStrategy}
import org.apache.flink.table.sources.tsextractors.{ExistingField, StreamRecordTimestamp, TimestampExtractor}
import org.apache.flink.table.sources.wmstrategies.{AscendingTimestamps, BoundedOutOfOrderTimestamps, PreserveWatermarks, WatermarkStrategy}

import scala.collection.mutable
import scala.collection.JavaConverters._

/**
  * Rowtime descriptor for describing an event time attribute in the schema.
  */
class Rowtime extends Descriptor {

  private var timestampExtractor: Option[TimestampExtractor] = None
  private var watermarkStrategy: Option[WatermarkStrategy] = None

  /**
    * Sets a built-in timestamp extractor that converts an existing [[Long]] or
    * [[Types.SQL_TIMESTAMP]] field into the rowtime attribute.
    *
    * @param fieldName The field to convert into a rowtime attribute.
    */
  def timestampsFromField(fieldName: String): Rowtime = {
    timestampExtractor = Some(new ExistingField(fieldName))
    this
  }

  /**
    * Sets a built-in timestamp extractor that converts the assigned timestamps from
    * a DataStream API record into the rowtime attribute and thus preserves the assigned
    * timestamps from the source.
    *
    * Note: This extractor only works in streaming environments.
    */
  def timestampsFromSource(): Rowtime = {
    timestampExtractor = Some(new StreamRecordTimestamp)
    this
  }

  /**
    * Sets a custom timestamp extractor to be used for the rowtime attribute.
    *
    * @param extractor The [[TimestampExtractor]] to extract the rowtime attribute
    *                  from the physical type.
    */
  def timestampsFromExtractor(extractor: TimestampExtractor): Rowtime = {
    timestampExtractor = Some(extractor)
    this
  }

  /**
    * Sets a built-in watermark strategy for ascending rowtime attributes.
    *
    * Emits a watermark of the maximum observed timestamp so far minus 1.
    * Rows that have a timestamp equal to the max timestamp are not late.
    */
  def watermarksPeriodicAscending(): Rowtime = {
    watermarkStrategy = Some(new AscendingTimestamps)
    this
  }

  /**
    * Sets a built-in watermark strategy for rowtime attributes which are out-of-order by a bounded
    * time interval.
    *
    * Emits watermarks which are the maximum observed timestamp minus the specified delay.
    *
    * @param delay delay in milliseconds
    */
  def watermarksPeriodicBounded(delay: Long): Rowtime = {
    watermarkStrategy = Some(new BoundedOutOfOrderTimestamps(delay))
    this
  }

  /**
    * Sets a built-in watermark strategy which indicates the watermarks should be preserved from the
    * underlying DataStream API and thus preserves the assigned watermarks from the source.
    */
  def watermarksFromSource(): Rowtime = {
    watermarkStrategy = Some(PreserveWatermarks.INSTANCE)
    this
  }

  /**
    * Sets a custom watermark strategy to be used for the rowtime attribute.
    */
  def watermarksFromStrategy(strategy: WatermarkStrategy): Rowtime = {
    watermarkStrategy = Some(strategy)
    this
  }

  /**
    * Internal method for properties conversion.
    */
  final override def addProperties(properties: DescriptorProperties): Unit = {
    val props = mutable.HashMap[String, String]()
    timestampExtractor.foreach(normalizeTimestampExtractor(_).foreach(e => props.put(e._1, e._2)))
    watermarkStrategy.foreach(normalizeWatermarkStrategy(_).foreach(e => props.put(e._1, e._2)))
    properties.putProperties(props.toMap.asJava)
  }
}

/**
  * Rowtime descriptor for describing an event time attribute in the schema.
  */
object Rowtime {

  /**
    * Rowtime descriptor for describing an event time attribute in the schema.
    */
  def apply(): Rowtime = new Rowtime()

}

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.table.descriptors;

import org.apache.flink.annotation.PublicEvolving;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.table.sources.tsextractors.TimestampExtractor;
import org.apache.flink.table.sources.wmstrategies.WatermarkStrategy;

import java.util.Map;

/** Rowtime descriptor for describing an event time attribute in the schema. */
@PublicEvolving
public class Rowtime implements Descriptor {

    private final DescriptorProperties internalProperties = new DescriptorProperties(true);

    // TODO: Put these fields into RowtimeValidator once it is also ported into table-common.
    // TODO: Because these fields have polluted this API class.
    public static final String ROWTIME = "rowtime";
    public static final String ROWTIME_TIMESTAMPS_TYPE = "rowtime.timestamps.type";
    public static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD = "from-field";
    public static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE = "from-source";
    public static final String ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM = "custom";
    public static final String ROWTIME_TIMESTAMPS_FROM = "rowtime.timestamps.from";
    public static final String ROWTIME_TIMESTAMPS_CLASS = "rowtime.timestamps.class";
    public static final String ROWTIME_TIMESTAMPS_SERIALIZED = "rowtime.timestamps.serialized";

    public static final String ROWTIME_WATERMARKS_TYPE = "rowtime.watermarks.type";
    public static final String ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_ASCENDING =
            "periodic-ascending";
    public static final String ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_BOUNDED = "periodic-bounded";
    public static final String ROWTIME_WATERMARKS_TYPE_VALUE_FROM_SOURCE = "from-source";
    public static final String ROWTIME_WATERMARKS_TYPE_VALUE_CUSTOM = "custom";
    public static final String ROWTIME_WATERMARKS_CLASS = "rowtime.watermarks.class";
    public static final String ROWTIME_WATERMARKS_SERIALIZED = "rowtime.watermarks.serialized";
    public static final String ROWTIME_WATERMARKS_DELAY = "rowtime.watermarks.delay";

    /**
     * Sets a built-in timestamp extractor that converts an existing {@link Long} or {@link
     * Types#SQL_TIMESTAMP} field into the rowtime attribute.
     *
     * @param fieldName The field to convert into a rowtime attribute.
     */
    public Rowtime timestampsFromField(String fieldName) {
        internalProperties.putString(
                ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD);
        internalProperties.putString(ROWTIME_TIMESTAMPS_FROM, fieldName);
        return this;
    }

    /**
     * Sets a built-in timestamp extractor that converts the assigned timestamps from a DataStream
     * API record into the rowtime attribute and thus preserves the assigned timestamps from the
     * source.
     *
     * <p>Note: This extractor only works in streaming environments.
     */
    public Rowtime timestampsFromSource() {
        internalProperties.putString(
                ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_SOURCE);
        return this;
    }

    /**
     * Sets a custom timestamp extractor to be used for the rowtime attribute.
     *
     * @param extractor The {@link TimestampExtractor} to extract the rowtime attribute from the
     *     physical type.
     */
    public Rowtime timestampsFromExtractor(TimestampExtractor extractor) {
        internalProperties.putProperties(extractor.toProperties());
        return this;
    }

    /**
     * Sets a built-in watermark strategy for ascending rowtime attributes.
     *
     * <p>Emits a watermark of the maximum observed timestamp so far minus 1. Rows that have a
     * timestamp equal to the max timestamp are not late.
     */
    public Rowtime watermarksPeriodicAscending() {
        internalProperties.putString(
                ROWTIME_WATERMARKS_TYPE, ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_ASCENDING);
        return this;
    }

    /**
     * Sets a built-in watermark strategy for rowtime attributes which are out-of-order by a bounded
     * time interval.
     *
     * <p>Emits watermarks which are the maximum observed timestamp minus the specified delay.
     *
     * @param delay delay in milliseconds
     */
    public Rowtime watermarksPeriodicBounded(long delay) {
        internalProperties.putString(
                ROWTIME_WATERMARKS_TYPE, ROWTIME_WATERMARKS_TYPE_VALUE_PERIODIC_BOUNDED);
        internalProperties.putLong(ROWTIME_WATERMARKS_DELAY, delay);
        return this;
    }

    /**
     * Sets a built-in watermark strategy which indicates the watermarks should be preserved from
     * the underlying DataStream API and thus preserves the assigned watermarks from the source.
     */
    public Rowtime watermarksFromSource() {
        internalProperties.putString(
                ROWTIME_WATERMARKS_TYPE, ROWTIME_WATERMARKS_TYPE_VALUE_FROM_SOURCE);
        return this;
    }

    /** Sets a custom watermark strategy to be used for the rowtime attribute. */
    public Rowtime watermarksFromStrategy(WatermarkStrategy strategy) {
        internalProperties.putProperties(strategy.toProperties());
        return this;
    }

    /** Converts this descriptor into a set of properties. */
    @Override
    public Map<String, String> toProperties() {
        final DescriptorProperties properties = new DescriptorProperties();
        properties.putProperties(internalProperties);
        return properties.asMap();
    }
}
